# **Bootcamp Ciencia de Datos e Inteligencia Artificial**
# **Módulo 1. Fundamentos de Data Science**
## **1. Programación en Python**

¡Te damos la bienvenida al Bootcamp de Data Science e Inteligencia Artificial!

Para iniciar este camino, es fundamental que sepas que este Bootcamp está diseñado especialmente para personas que tienen un gran interés y motivación para dominar las herramientas necesarias para desenvolverse en el campo laboral de la ciencia de datos. Esperamos que cada escalón sea disfrutable y te permita aprender a resolver cualquier problemática de analíticos avanzados.

El objetivo de esta semana es comprender la lógica de la programación en Python, a fin de usar sus bases de la mejor manera durante las siguientes semanas.

¡Comencemos!

## *1.1 Lenguaje Python*

**Python** es un lenguaje de programación que se ocupa en una gran variedad de áreas, debido a su versatilidad y al soporte que tiene por parte de una comunidad de programadores de todo el mundo; además, si se comprende su lógica de programación, es más sencillo de aprender que otros lenguajes.

Si quieres revisar algunas de las ventajas de este lenguaje, consulta el artículo *How Python became the language for data science*, publicado en el Data Science Central: (https://www.datasciencecentral.com/how-python-became-the-language-for-data-science/)

A Python se le deben dar instrucciones de operación, las cuales, en los lenguajes de programación, se llaman *declaraciones*.

Veamos un ejemplo de cómo funciona el lenguaje de Python. Usaremos una declaración muy sencilla, llamada `print`, la cual muestra en la pantalla un texto. Observa lo siguiente:

In [5]:
print('Hola Mundo')

Hola Mundo


Si ejecutas la línea de arriba, verás en la pantalla `Hola mundo`. La estructura de esa declaración se puede analizar paso a paso.

1. La palabra `print` al inicio.
2. Inmediatamente después de la palabra `print`, sin dejar espacios, se coloca un paréntesis de apertura `(`.
3. Ya sea entre comillas dobles (`“texto"`) o simples (`‘texto'`), se ingresa un texto que, en Python, es conocido como **cadena**, **cadena de texto** o **string**.
4. Por último, se coloca un paréntesis de cierre `)`.

Al respecto, es importante definir una de las primeras y más importantes reglas de Python: por cada paréntesis de apertura `(` debe haber uno de cierre `)`, es decir, jamás deberá faltar alguno de los dos.

Ahora que iniciamos con Python, veremos los distintos tipos de datos.

##### Ponte a prueba
Escribe un saludo usando la declaración `print`; es muy sencillo.

In [6]:
print ('Hola Fati')

Hola Fati


## *1.2 Datos Numéricos*
En **Python**, así como en otros lenguajes de programación, los números permiten hacer operaciones matemáticas y funciones muy específicas por la naturaleza de este tipo de información, así como distintos "tipos de números". A continuación, podremos ver algunas cosas que es posible hacer con estos datos y algunas de sus características.


### 1.2.1 Operaciones básicas

En **Python**, es posible hacer operaciones sencillas; ejemplo de ello es que puedes hacer programas como los siguientes:

In [7]:
print(3 + 5)
print(8 - 2)
print(7 + 5 - 3)

8
6
9


Con base en lo anterior, veremos que el signo `+` es interpretado por Python como una suma, y el signo `-`, como una resta. Además, se puede multiplicar con el caracter `*` y dividir con la barra `/`; ejemplo de lo anterior son los siguientes códigos:

In [8]:
print(5 * 6)
print(4 + (3 * 5))
print(12 / 4)

30
19
3.0


Quizá te preguntes ¿por qué el último resultado fue `3.0` y no solo `3`? Esto se debe a que en Python las divisiones siempre son flotantes. A continuación se explicarán los dos «tipos de números» que existen:

> Los números **enteros** o **integers** son los que se usan con más frecuencia en el lenguaje de la programación. Como su nombre lo indica, son números enteros, es decir, se ignora cualquier valor que aparece después de un punto decimal. Ejemplos: 4, 8, 0, -1.
>
> Los números **flotantes** o **float** se consideran decimales. Ejemplos: 0.5, 10.75, 3.74925762, 2.0.

En ambos casos es común usar sus nombres en inglés.

No solo la división genera un número flotante como consecuencia de usar un número entero sobre otro entero, sino también la suma, la resta y la multiplicación, pues, por lo menos, uno de esos números es flotante también.

Veamos un ejemplo de lo anterior:

In [9]:
print(5.0 + 4)
print(10.0 * 3.0)
print(10 * 3)
print(15 / 5)

9.0
30.0
30
3.0


- En el primer ejemplo, debido a que se sumó un **flotante** y un **entero**, el resultado es un **flotante**.
- En el segundo ejemplo, se multiplican dos **flotantes**, por lo que el resultado es un **flotante**.
- En el tercer ejemplo, solo se están multiplicando **enteros**, y debido a que ninguno es **flotante**, el resultado seguirá siendo un **entero**.
- En el último ejemplo, vemos una división de dos **enteros**, en la que ninguno es **flotante**; sin embargo, el resultado es un número **flotante**, ya que las divisiones siempre arrojarán ese tipo de datos.

### 1.2.2 Exponentes y raíces cuadradas y cúbicas

En **Python** no solo se puede sumar, restar, multiplicar y dividir, también se pueden colocar números como exponentes.

Esto se hace con el comando doble asterisco (`**`), y, al igual que en la división, siempre regresa un valor **flotante**.

In [10]:
print(4**3)

64


Lo anterior es lo mismo que indicar $4^3=64$, esto permitiría una torre de exponentes como por ejemplo $2^{3^2}$ El cual deberemos indicar como:

In [11]:
print(2**3**2)

512


Por último existe un pequeño truco que nos permitiría conocer raíces cuadras o raíces cúbicas recordando algunas clases de álgebra:
$$
\sqrt[2]{25} = 25^{\frac{1}{2}} \text{ Raiz cuadrada es lo mismo que }  n \text{ elevado a   } \frac{1}{2} \\
\sqrt[3]{64} = 25^{\frac{1}{3}} \text{ Raiz cúbica es lo mismo que }  n \text{ elevado a   } \frac{1}{3}
$$
Por lo que la raíz cuadrada y la raíz cúbica se expresarían así:

In [12]:
print(25 ** (1/2))
print(64 ** (1/3))

5.0
3.9999999999999996


### 1.2.3 Cociente y residuo

En las operaciones básicas, hay dos comandos que son muy útiles en la programación, y ambos están relacionados con las divisiones.

El **cociente** es el número entero que resulta de una división en la que no se consideran los valores decimales; por ejemplo, si hacemos la división $\frac{22}{6}$, sabemos que el resultado es $3.66666$, pero cuando pedimos el cociente, el resultado ignorará los decimales, incluso si es cercano a un número entero.

> **NOTA**: No debes confundir cociente con redondear, son valores totalmente distintos.

En **Python**, el **cociente** se pide con la expresión doble barra (`//`), por lo tanto, la expresión para encontrar el cociente de $\frac{22}{6}$ es:

In [13]:
print(22 // 6)

3


Observando detenidamente vemos que el resultado de pedir un cociente es un **entero**, aunque si alguno de los dos fuera un **flotante** el resultado será un **flotante** también (similar a lo que sucede con la suma y la resta).

¿Que sucedería si requerimos el residuo de una división? Recordemos que el residuo es aquel número que sobra dentro de una división, lo explicaremos con un ejemplo sencillo.

Tenemos 22 dulces, los cuales repartiremos entre 6 niños. Si hacemos la división a cada niño le tocarían 3 dulces, pero sobrarían 4 dulces.
$$
22 - (6 \times 3) = 22 - 18 = 4
$$
Esos 4 dulces sobrantes, los cuales no repartiremos porque no alcanzaría para dar a cada niño, es el residuo.

Tanto el **cociente**, como el **residuo** están presentes en todas las divisiones, veamos algunos ejemplos:
$\frac{25}{4}$, cociente = $6$, residuo = $1$
$\frac{43}{8}$, cociente = $5$, residuo = $3$
$\frac{15}{5}$, cociente = $3$, residuo = $0$

El **residuo** se solicita en **Python** con el símbolo de porcentajes (`%`), ejemplos de ello son:

In [14]:
print((55 % 7) * 70)

420


Como podrás ver por lo general los resultados son **enteros**, sólo en algunos casos en que el residuo sea con decimales este será un **flotante**.

 A manera de recordatorio estas son las operaciones básicas que se pueden hacer con los números:

| Operación      |  Símbolo  |   Ejemplo   | Tipo de salida   |
| -------------- | :-------: | :---------: | :--------------- |
| Suma           |    `+`    |  `15 + 18`  | entero, flotante |
| Resta          |    `-`    |  `33 - 27`  | entero, flotante |
| Multiplicación |    `*`    |  `12 * 3`   | entero, flotante |
| División       |    `/`    |  `24 / 5`   | flotante         |
| Exponenciación |   `**`    |   `5**3`    | entero, flotante |
| Raíz           | `**(1/2)` | `81**(1/2)` | flotante         |
| Cociente       |   `//`    |   `38//6`   | entero, flotante |
| Residuo        |    `%`    |  `50 % 4`   | entero, flotante |



**Ponte a prueba**

Usando las funciones que hemos aprendido resuelve el siguiente cuestionamiento:

En un almacén se tienen 55 pares de pantalones, los cuales se deberán agrupar en 7 paquetes, y las piezas sobrantes se podrán vender de forma individual a $70.

1. **¿Cuántos pantalones habrá en cada paquete? ** Habrán 7 pares de pantalones en cada paquete.
2. **¿Cuál es el monto total que se percibirá de los pantalones vendidos individualmente?** 420 pesos por el total de los 6 pantalones que sobraron.
3. Usando `print` haz que estos resultados se muestren en pantalla.
print((55%7)*70)



In [15]:
pantalones = 55
paquetes = 7
precio_individual = 70
print(pantalones//paquetes)
print((pantalones % paquetes)* precio_individual)

7
420


## 1.3 Cadenas de texto

Al inicio aplicamos, por primera vez, la declaración `print`, que permite mostrar un texto o una operación matemática en la pantalla. Ahora bien, se mencionó que todo texto debe escribirse entre comillas (ya sean dobles `“texto”` o simples `'texto'`).

En la programación, no se le llama *texto*, sino **cadena**, **cadena de texto** o **string**, como le llamaremos de aquí en adelante.

Recordemos la función `print`.

In [16]:
print('¡Hola, mundo!')
print("¡Hola, mundo!")

¡Hola, mundo!
¡Hola, mundo!


### 1.3.1 Caracteres especiales

Vale la pena indicar que hay algunos caracteres que no se pueden imprimir, por ejemplo, los apóstrofos. Para ello, simplemente se coloca una barra invertida (`\`) antes de esos caracteres.

In [17]:
print('I\'m glad be with you')

I'm glad be with you


En el caso de que queramos usar apóstrofos como en el ejemplo anterior, podemos usar dobles comillas para nuestra **cadena de texto**.

In [18]:
print("I\'m glad be with you")

I'm glad be with you


### 1.3.2 Saltos de línea

Quizá existan algunos casos en los que requeriremos más de una línea de texto. Para ello, existen dos formas de hacerlo.

La primera es que en la **cadena de texto** coloquemos `\n`, que quiere decir que queremos una nueva línea.

In [19]:
print('Hola, buenos días\nmi nombre es José.\nGusto en conocerte.')

Hola, buenos días
mi nombre es José.
Gusto en conocerte.


Otra forma de hacerlo es que la **cadena de texto** comience con tres comillas(`‘'’`) o doble comillas (`“"”`) y termine igualmente con tres comillas(`‘'’`) o doble comillas (`“"”`).

In [20]:
print('''
Hola, buenos días
mi nombre es José.
Gusto en conocerte.
''')

print("""
Hola, buenos días
mi nombre es José.
Gusto en conocerte.
""")


Hola, buenos días
mi nombre es José.
Gusto en conocerte.


Hola, buenos días
mi nombre es José.
Gusto en conocerte.



### 1.3.3 Operaciones con cadenas de texto

#### Concatenar

Cuando tenemos dos **cadenas de texto**, podemos juntarlas en una sola; a eso se le llama **concatenar**. Para ello, usaremos la suma (`+`), como se muestra a continuación:

In [21]:
print('Buenas tardes, ' + '¿Cómo estás?')

Buenas tardes, ¿Cómo estás?


Vale la pena indicar que no podemos concatenar una **cadena de texto** con un **entero** o un **flotante**, por ello es importante que mencionemos que `2` no es lo mismo que `'2'`.
- `2` sin comillas es un **entero** y este elemento se puede sumar, restar, dividir, étc.
- `'2'`con comillas es una **cadena de texto**, por lo que se considera un texto, no puede sumarse, restarse, étc.

Por lo que para concatenar un número deberemos hacerlo de la siguiente manera:

In [22]:
print('Quiero ' + '2' + ' boletos.')

Quiero 2 boletos.


Multiplicar cadenas

Aunque arriba indicamos que una **cadena de texto** no se puede sumar, restar o dividir con un número ya sea **entero** o **flotante**, en **Python** si es posible multiplicar, lo que dará como resultado la **cadena de texto** repetida por la cantidad por la que se multiplicó, por ejemplo:

In [23]:
print('Hola!' * 4)

Hola!Hola!Hola!Hola!


Para que un número se repita varias veces es necesario considerarlo como **cadena de texto** y usar el mismo método:

In [24]:
print('8' * 3)

888


> **NOTA**: El resultado de multiplicar una **cadena de texto** nos dará como resultado otra **cadena de texto** por lo que aplican las mismas restricciones de poder sumar, restar o dividir.

Como se ha indicado en la nota, el `‘888’` que hemos obtenido de resultado de nuestro ejemplo anterior es una **cadena de texto** por lo que no es posible hacer operaciones aritméticas con ese valor.

##### Ponte a prueba

Investiga un poema pequeño y usando `print` haz que se muestre en pantalla, al finalizar el poema deberás hacer que se vea una línea donde indica que has finalizado, como esta:
```
Poema línea 1,
poema línea 2,
poema lìnea 3.
====================================
```

Recuerda hacer punto y aparte en el poema que escribas e intenta que la línea final sea generada multiplicando una **cadena de texto** con un número.

In [25]:
print('''
    Su mirada, cansada de ver pasar
    las rejas, ya no retiene nada más.
    Cree que el mundo está hecho,
    de miles de rejas y, más allá, la nada.
    ''')


    Su mirada, cansada de ver pasar
    las rejas, ya no retiene nada más.
    Cree que el mundo está hecho,
    de miles de rejas y, más allá, la nada.
    


### 1.3.4 Operadores booleanos

Los **booleanos** son datos que solo pueden tener dos estados: **verdadero** o **falso**. Estos operadores son usados en comparaciones o en pruebas lógicas, funciones que se explicarán más adelante. Por ahora, indicaremos que en Python son escritos como `True` para valores verdaderos (con la letra `T` mayúscula y el resto de los caracteres en minúsculas) y como `False` para valores falsos (con la letra `F` mayúscula y el resto de los caracteres en minúsculas).

In [26]:
True
False

False

### 1.3.5 Variables

En Python, los datos se pueden almacenar en elementos llamados **variables**, lo cual permite usar ese valor almacenado más adelante, en el programa, y ofrece más dinamismo al momento de ejecutar un código.

Es muy sencillo asignar una variable: basta que indiquemos, primero, el nombre de nuestra variable; inmediatamente después, en la misma línea, escribimos un signo de igual `=`, y, posteriormente, indicamos el valor que queremos guardar.

Ejemplos:

In [27]:
nombre = 'María'
edad = 28
estatura = 1.62

Si ingresamos y corremos el ejemplo de arriba no nos devolverá nada, ya que sólo hemos pedido «guardar» la información en nuestras **variables**, no otra operación más.

Como podrás ver en el ejemplo de arriba las **variables** pueden almacenar cualquier tipo de dato y a diferencia de otros lenguajes de programación no es necesario indicar qué tipo de dato es, ya que **Python** lo interpreta inmediatamente, sólo recuerda que las **cadenas de texto** siempre se escriben entre comillas simples (`‘texto'`) o dobles (`“texto"`).

Aunque puedes indicar casi cualquier nombre que quieras a tus **variables** existen una serie de restricciones:

Si ingresamos y corremos el ejemplo de arriba no nos devolverá nada, ya que sólo hemos pedido «guardar» la información en nuestras **variables**, no otra operación más.

Como podrás ver en el ejemplo de arriba las **variables** pueden almacenar cualquier tipo de dato y a diferencia de otros lenguajes de programación no es necesario indicar qué tipo de dato es, ya que **Python** lo interpreta inmediatamente, sólo recuerda que las **cadenas de texto** siempre se escriben entre comillas simples (`‘texto'`) o dobles (`“texto"`).

Aunque puedes indicar casi cualquier nombre que quieras a tus **variables** existen una serie de restricciones:

- Sólo se permiten usar letras, números y guiones bajos (`_`).
- El nombre de la **variable** no puede iniciar con un número.
- No puede incluir espacios, para ese caso se aconseja el uso del guión bajo (`_`).
- No se pueden usar palabras reservadas (las aprenderán a lo largo del curso).

  > **NOTA**: **Python** es sensible a mayúsculas y minúsculas, eso quiere decir que no es lo mismo `Grupo` que `grupo`, o incluso qué `GRUPO`.

¿Para qué sirve el uso de las **variables**? Cómo se indicó previamente permiten que la información almacenada sea utilizada más adelante.

Para comprender mejor ello usaremos las tres **variables** que indicamos previamente: `nombre`, `edad` y `estatura`. Es importante que se hayan definido las **variables**, de lo contrario **Python** no reconocerá lo que pedimos almacenar.

A continuación vamos a hacer una serie de operaciones con los valores almacenados, es importante recordar las restricciones que tienen cada tipo de datos (ejemplo: las **cadenas de texto** no se pueden restar o dividir), ya que pueden generar un error.

Para invocar una **variable** sólo bastará con escribir el nombre de la misma como lo definimos previamente. Si has sido observador notarás porque es importante que una **cadena de texto** esté entrecomillada, ya que usualmente un texto sin comillas en realidad hace referencia a una **variable**, ten en mente siempre esto para cuando realices tus proyectos.

In [28]:
print('Buenos días ' + nombre)
print(edad - 18)
print(estatura / 2)

Buenos días María
10
0.81


Como podrás ver es como si al segundo código hubiéramos pegado directamente lo que almacenamos en las **variables**, esto permite que a las ***variables*** podamos usarlas una infinidad de veces.

In [29]:
print(estatura / 2)
print(estatura / 3)
print(estatura / 4)
print(estatura / 5)
print(estatura / 6)

0.81
0.54
0.405
0.324
0.27


Si quisiéramos cambiar o actualizar la información en una variable sólo basta con hacer el mismo proceso como cuando se definió la **variable** por primera vez:

In [30]:
edad = '28'
print('Edad de Ana el año pasado: ' + edad)
edad = '29'
print('Edad de Ana este año: ' + edad)

Edad de Ana el año pasado: 28
Edad de Ana este año: 29


## 1.4 Estructura de datos

### Listas

Una forma de almacenar una serie de elementos es mediante las **listas**. Estos se identifican fácilmente, ya que los valores están entre corchetes (`[elemento0, elemento1, elemento2]`) y sus valores se separan con comas (`,`). Ejemplos de listas son las siguientes:

In [31]:
primera_lista = [3, 8, 12, 25]
segunda_lista = [1.4, 3.0, 2.75, 8.00]
tercera_lista = ['Alberto', 'Benito', 'César']
cuarta_lista = [3, 'Hola', 2.25]

Como puedes ver, la lista se definió como las variables. Sus elementos pueden ser **enteros**, como se muestra en la `primera_lista`; **flotantes**, como en la `segunda_lista`; **cadenas de texto**, como en la `tercera_lista`, o una mezcla, en la que cada elemento puede ser cualquier tipo de dato, como sucede en la `cuarta lista`.

Para ver los elementos, basta con invocarlos, como lo hacemos con una **variable**:

In [32]:
print(primera_lista)
print(segunda_lista)

[3, 8, 12, 25]
[1.4, 3.0, 2.75, 8.0]


Un aspecto que tiene la **lista** es que podemos indicarle que nos regrese un elemento específico de todos los que la componen. Por ejemplo, la `primera_lista` está compuesto por `[3, 8, 12, 25]`. Si quiero que me indique el primer valor, solo hace falta mencionar su **índice** entre corchetes `[índice]` e inmediatamente después del nombre de la lista, como se muestra a continuación:

In [33]:
print(primera_lista[0])

3


Para el resto, sería:

In [34]:
print(primera_lista[1])
print(primera_lista[2])
print(primera_lista[3])

8
12
25


> **NOTA**: En **Python**, el primer índice de una lista comienza en `0`, por lo que el índice del segundo elemento es `1`, el del tercero es `2`, el del cuarto es `3`, y así sucesivamente. Recuerda muy bien esto, ya que está presente en todo el curso y es uno de los aspectos más importantes de este lenguaje de programación.

Por raro que parezca, pueden haber listas de listas. Por lo general, cuando hacemos una lista, cada elemento solo contiene un valor, pero es posible que uno o varios de esos valores sean otra lista, como se muestra en el ejemplo.

In [35]:
lista_ejemplo = [2, 4, [6, 7, 8]]

En el ejemplo anterior `lista_ejemplo`, tienes tres elementos, que desglosaremos con el siguiente código:

In [36]:
print(lista_ejemplo[0])
print(lista_ejemplo[1])
print(lista_ejemplo[2])

2
4
[6, 7, 8]


Como puedes ver, el primer elemento (índice `0`) es el **entero** 2, el segundo (índice `1`) es el **entero** 4 y el tercero (índice `2`) es la lista `[6, 7, 8]`.

Si quisiéramos acceder a un valor específico de la lista dentro de la lista, entonces deberemos indicarlo con un «índice después del índice», el cual puedes escribir de la siguiente manera:

In [37]:
print(lista_ejemplo[2][0])
print(lista_ejemplo[2][1])
print(lista_ejemplo[2][2])

6
7
8


En `lista_ejemplo`, la primera línea del códigose interpreta de la siguiente manera: de su tercer elemento (índice `2`) nos regresará el primer valor (índice `0`).

Solo para tenerlo en consideración: es posible hacer listas vacías, las cuales se llenarán posteriormente, y aunque en este momento parece no tener utilidad, sí la tendrá a lo largo del curso.

In [38]:
lista_vacia = []

#### 1.4.1Cadena de texto como una lista

Una vez familiarizados con el código para invocar los elementos de una lista, es posible que **Python** considere una **cadena de texto** como si se tratara de una lista, en la que el primer carácter sería el elemento del primer índice (`0`), el segundo carácter sería el segundo índice (`1`), y así sucesivamente.

Puedes verlo en el siguiente ejemplo:

In [39]:
mensaje = 'Feliz cumpleaños'
mensaje[3]

'i'

La siguiente estructura nos indicaría qué índice tiene cada carácter.
```
           0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15
mensaje = 'F  e  l  i  z     c  u  m  p  l  e  a  ñ  o  s'
```

#### 1.4.2 Índices negativos

Otra forma de invocar los elementos es que, en lugar de que nos dé los valores de izquierda a derecha, lo haga de derecha a izquierda, o sea, que considere primero el último valor. Esto se logra indicando con valores negativos, donde `-1` es el último elemento en la **lista**, `-2` es el penúltimo, `-3` es el antepenúltimo y así sucesivamente. Cuando se buscan valores de derecha a izquierda, se comienza con el valor `-1`, pues no existe el valor `-0`. Sin embargo, cuando es de izquierda a derecha, sí se inicia con `0`.

Veamos un ejemplo similar al que hicimos, con un mensaje de «Feliz cumpleaños». Vale la pena indicar que esto es permitido en cualquier tipo de **lista**.

In [40]:
mensaje = 'Feliz cumpleaños'
mensaje[-3]

'ñ'

Podemos entender esto con la siguiente estructura:

```
           -16 -15 -14 -13 -12 -11 -10  -9  -8  -7  -6  -5  -4  -3  -2  -1
mensaje = '  F   e   l   i   z       c   u   m   p   l   e   a   ñ   o   s'
```

#### 1.4.3 Actualizar un elemento de una lista

Es común que uno o varios elementos de una **lista** deban modificarse en el código. Para ello, se deberá hacer algo similar a lo que se hizo con las variables: solo indicamos el elemento a cambiar e indicaremos el nuevo valor.

In [41]:
compras = ['leche', 'cereal', 'atún', 'arroz']
print(compras)
compras[2] = 'pollo'
print(compras)

['leche', 'cereal', 'atún', 'arroz']
['leche', 'cereal', 'pollo', 'arroz']


En la primera línea de código hemos creado la lista `compras` con cuatro elementos, en la segunda línea le hemos dicho a **Python** que imprima la lista que generamos previamente. Para la tercera línea hemos pedido que el elemento en el índice `2`(el tercer valor, ya que recordemos que de izquierda a derecha inicia en `0`) tenga un nuevo valor llamado `pollo`, es por ello que en la cuarta línea hemos vuelto a pedir que nos imprima la lista y el elemento cambiado se puede observar.

#### 1.4.4 Operaciones con listas

Las listas tienen algunas características que pueden asemejarse a las **cadenas de texto**, como es el caso de concatenar datos. Para ello, en una lista que hayamos definido, solo indicaremos, con un signo de suma (`+`), que queremos agregar más datos, los cuales deberán estar entre corchetes (`[elemento1, elemento2, elemento3]`).

Un ejemplo de lo anterior:

In [42]:
numeros_pares = [2, 4, 6, 8]
print(numeros_pares)
print(numeros_pares + [10, 12, 14])

[2, 4, 6, 8]
[2, 4, 6, 8, 10, 12, 14]


Como se puede ver en el ejemplo de arriba la primera lista es la que definimos en la variable `numeros_pares` en la lista posterior hemos concatenado a la lista original tres datos más. Hay que tener especial cuidado en que los valores agregados se han ingresado con elementos entre corchetes.

De igual forma que con la **cadena de texto** podemos repetir los datos con el operador para multiplicar, esto es con el asterisco (`*`).

In [43]:
numeros_impares = [1, 3, 5, 7]
print(numeros_pares)
print(numeros_pares * 4)

[2, 4, 6, 8]
[2, 4, 6, 8, 2, 4, 6, 8, 2, 4, 6, 8, 2, 4, 6, 8]


La primera **lista** que podemos observar en el ejemplo anterior son los elementos que definimos en `numeros_impares`, para que posteriormente la **lista** de abajo muestre los elementos en la **lista** cuatro veces, ya que eso se ha definido al multiplicar nuestra **lista** por cuatro.

#### 1.4.5 Longitud

Hasta el momento hemos aprendido la declaración `print`. En esta ocasión, aprenderemos dos más. La primera de ellas nos indica cuántos elementos hay en una **lista**, la cual es llamada `len` (derivada de la palabra *length*, que se traduce como *longitud*) y se usa de la siguiente manera:

In [44]:
lista_ejemplo = [3, 6, 9, 12, 15, 18]
print(len(lista_ejemplo))

6


Como podrás ver la declaración `len`es inmediatamente seguido por el nombre de la lista entre paréntesis (`(lista)`). Esta función indica la cantidad de elementos no importando qué tipo de datos contenga.

#### 1.4.6 _In_ y _not in_

Ahora podrás utilizar los **operadores booleanos**, o sea, `True` y `False`. Lo que haremos a continuación es preguntar a **Python** si un elemento está dentro de una lista. Veamos un ejemplo:

Imagina que tienes una **lista** con el nombre de varios alumnos y quieres saber si en esa lista está incluida `Alejandra` y `Mauricio`. Si ellos están en la lista, **Python** dirá `True`; si no están, `False`.

Lo primero que haremos es la **lista**.

In [45]:
alumnos = ['Armando', 'Bruno', 'Claudia', 'Daniela', 'Alejandra']

Ahora preguntaremos con la palabra reservada `in` si `Alejandra` está en la lista que previamente generamos. No olvidemos usar `print` para que el resultado se vea en pantalla:

In [46]:
print('Alejandra' in alumnos)

True


Esto se interpreta como «Es **verdadero** que el elemento `Alejandra` está en la lista `alumnos`.»

Ahora preguntaremos si `Mauricio` está está en la misma lista:

In [47]:
print('Mauricio' in alumnos)

False


Esto se interpreta como «Es **falso** que el elemento `Mauricio` está en la lista `alumnos`.»

Por otro lado podemos preguntar si es o no cierto que un elemento **no** está dentro de una lista, veamos un ejemplo con los mismos nombre que usamos en los ejemplos anteriores.

In [48]:
print('Alejandra' not in alumnos)

False


Esto se interpreta como «Es **falso** que el elemento `Alejandra` **no está** en la lista `alumnos`.»

Por otro lado usando el código:

In [49]:
print('Mauricio' not in alumnos)

True


El cual tendrá una interpretación como «Es **verdadero** que el elemento `Mauricio` **no está** en la lista `alumnos`.»

##### Ponte a prueba
Genera una lista llamada `numeros_primos` que contenga los siguientes valores:
1
2
3
5
7
11
13
17
19
23
29
31

Ahora deberás indicar cuál es el tercer valor, el séptimo valor y el último valor usando las expresiones que aprendimos en este apartado.

In [51]:
numeros_primos = [1, 2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31]
print(numeros_primos[3])
print(numeros_primos[7])
print(numeros_primos[-1])

5
17
31


### 1.4.7 Tuplas

Ahora veremos un concepto muy similar al de **lista**, pero que tiene características distintas: **tuplas**.

Una **tupla**, a diferencia de una **lista**, es **inmutable**, lo cual quiere decir que no se puede cambiar después de ser declarada. La forma de diferenciar una de otra es que, como ya sabemos, las **listas** son una serie de elementos que van entre corchetes (`[elemento_1, elemento_2, elemento_3]`), y las **tuplas** van entre paréntesis (`(elemento_1, elemento_2, elemento_3)`).

Al igual que en las **listas**, para acceder a sus elementos se invocan mediante el índice del elemento que se quiere consultar. Ejemplo:

In [ ]:
frutas = ('manzana', 'melón', 'sandía', 'mango')
print(frutas[2])

sandía


> **NOTA**: Recuerda que los índices en **Python** inician en `0`, por lo que `0` indica el primer elemento, `1` el segundo elemento y así sucesivamente.

Como se Indicó previamente las **tuplas** son **inmutables** y tratar de modificar los valores como lo hacemos con una **lista** nos resultará en un error:

```
frutas = ('manzana', 'melón', 'sandía', 'mango')
print(frutas[1]) = 'pera

File "<ipython-input-49-f401e5d336cf>", line 2
    print(frutas[1]) = 'pera'
                             ^
SyntaxError: can't assign to function call
```

### 1.4.8 Diccionario

Los **diccionarios** son estructuras que permiten asignar claves a los valores, como si a una lista, en lugar de tener los índices clásicos (como 0, 1, 2, 3, etcétera), nosotros le asignaramos la que nos guste, por ejemplo, `'A'`, `'B'`, `'C'`, `‘México’`, `‘Alejandro’`, etcétera.

Su estructura es similar a la de las **listas**, pero con algunos cambios, sobre todo porque, en lugar de corchetes (`[]`), lleva llaves (`{}`).

La estructura que emplea es la siguiente:

In [ ]:
ejemplo = {'clave_1': 'elemento_1', 'clave_2': 'elemento_2', 'clave_3': 'elemento_3'}

Vamos a entender lo que tenemos arriba, a un **diccionario** lo hemos llamado `ejemplo`, donde contiene tres elementos, verás que están separados por comas (`,`).

El primer elemento tiene como clave la palabra `'clave_1'` y el elemento se llama `'elemento_1'`, si esto fuera una **lista** es como si aquí estuviera el `'elemento_1'` y que nosotros para referenciarlo usáramos el índice `0` (de que es el primer elemento), pero como esto no es una **lista** y es un **diccionario**, no usa índices y haremos referencia a el elemento con la clave que le asignamos, o sea `clave_1`. Esto se entenderá mejor con un ejemplo:

In [ ]:
ejemplo = {'clave_1': 'elemento_1', 'clave_2': 'elemento_2', 'clave_3': 'elemento_3'}
print(ejemplo['clave_1'])
print(ejemplo['clave_3'])

elemento_1
elemento_3


## 1.5 Estructuras de control

Una vez comprendidos los **operadores booleanos**, podemos analizar algunos aspectos en los que son muy útiles.

### Comparaciones

Las comparaciones son sentencias lógicas que solo pueden arrojar dos posibles resultados: verdadero (`True`) o falso (`Falso`). En nuestro idioma natural, así como en las matemáticas, es fácil comprender esto; es como preguntar si 20 es mayor que 15, o si la palabra *elección* es lo mismo que *selección* (no desde la perspectiva gramatical, sino, literalmente, letra por letra, incluyendo tildes, espacios y letras mayúsculas).

Para llevar a cabo ese análisis, es muy útil recordar los caracteres que se utilizan para comparar dos elementos:

| Caracteres | Significado | Ejemplo | Interpretación |
| :-----: | ----- | :-----: | ----- |
|  `==`   | idéntico a        | `a == b` | `a` es idéntico a `b` |
|  `!=`   | distinto a        | `a != b` | `a` es distinto a `b` |
|   `>`   | mayor que         | `a > b`  | `a` es mayor que `b` |
|  `>=`   | mayor o igual que | `a >= b` | `a` es mayor o igual que `b` |
|   `<`   | menor que         | `a < b`  | `a` es menor que `b` |
|  `<=`   | menor o igual que | `a <= b` | `a` es menor o igual que `b` |

> **NOTA**: Es importante no confundir `=`con `==`, ya que no representan lo mismo: mientras `=`sirve para definir **variables** o **listas**, `==` permite comparar si dos elementos son idénticos.

Recuerda que `True` quiere decir que la condición es cierta, y `False`, que es falsa. Con eso en mente, podemos ver algunos ejemplos:

In [ ]:
print(8.0 == 8)
print('Archivo' != 'archivo')
print(6 > 6)
print(8.0 >= 8.0)
print(3**3 < 30)
print(64**(1/2) <= 28//5)

True
True
False
True
True
False


Revisaremos, línea por línea, cada resultado:

- En la primera línea, se hace la comparación `8.0 == 8`, la cual se interpreta como «¿8.0 es igual a 8?». A pesar que un número es **entero** y el otro **flotante**, los dos valores son iguales; por lo tanto, el resultado es `True`.
- En la segunda línea se hace la comparación `'Archivo' != 'archivo'`, la cual se interpreta como «¿La **cadena de texto** ‘Archivo’ es distinto a la **cadena de texto** ‘Archivo’?». A pesar que ambas **cadenas de texto** son la misma palabra, una fue escrita con mayúscula y la otra no, por lo que _no son iguales_ y el resultado es `True` (la pregunta es si son distintos).
- En la tercera línea se hace la comparación `6 > 6`, la cual se interpreta como «¿6 es mayor que 6?». Ambos valores son iguales o, en otras palabras, uno no es mayor que otro; por lo tanto, la respuesta es `False`.
- En la cuarta línea se hace la comparación `8.0 >= 8.0`, la cual se interpreta como «¿8.0 es mayor o igual que 8.0?». Aunque un valor no es mayor que otro, la pregunta no solo cuestiona si es mayor, sino también si los valores son iguales; por ende, la respuesta es `True`.
- En la quinta línea se hace la comparación `3**3 < 30`, la cual se interpreta como «¿3 al cubo es menor que 30?». Para ello, desglosamos `3**3`, que arrojará 27. En ese sentido, la pregunta se reinterpretaría como «¿27 es menor que 30?», debido a que la respuesta es sí; entonces, **Python** nos arrojará `True`.
- En la sexta línea se hace la comparación `64**(1/2) <= 28//5`, la cual se interpreta como «¿la raíz cuadrada de 64 es menor o igual que el cociente de 28 entre 5?». Para ello, desglosamos `64**(1/2)`, el cual sería 8 (la raíz cuadrada de 64); también desglosamos `28//5`, el cual arrojará como respuesta 5 (28 entre 5 son 5, sin considerar ningún decimal), por lo que la pregunta se reinterpreta como «¿8 es menor o igual que 5?», debido a que la respuesta es no; entonces, **Python** nos indicará `False`.

#### 1.5.1 Declaración _if_

Una vez comprendida la estructura de una condición podemos usarlas para que **Python** realice una acción dado que se cumpla una petición. Para ello analizaremos la declaración `if`, que en inglés significa «si» (condicional), no confundir con sí (afirmación).

La estructura para esta declaración es la siguiente:
```
if a == b:
    print(a + b)
```

Veamos parte por parte:

- Primero se escribe la palabra reservada `if`, recuerda que una palabra reservada en **Python** quiere decir que no debe ser usada para nombrar una variable o una lista.

- Después de `if` indicamos la condición a cumplir, en el ejemplo se pregunta si `a` y `b` son idénticos, pero puede emplearse cualquier otra comparación. Recuerda que si preguntarás si dos valores son iguales uses `==`, y evites `=` que generará error en tu código.

- Inmediatamente después, en la misma línea después de tu argumento de comparación debes escribir dos puntos (`:`), esto es importante, ya que si no lo haces **Python** no comprenderá qué quieres hacer.

- Para pasar a la siguiente línea de código sólo basta con presionar la tecla «Enter». Si estás escribiendo tu código en un _software_ para procesamiento de lenguajes de programación o IDLE, verás que al hacer presionado «Enter», no solo pasó a la siguiente línea, sino que el cursor se ha movido un poco a la derecha, a esto se le conoce como **indentación**, es una serie de espacios que indican que esa línea que estás escribiendo es parte de lo que está en la línea anterior, la **indentación** como se ha comentado antes por lo general es colocada de forma automática, aunque puedes hacer lo mismo con la tecla «Tabular» presionada una sola vez o con cuatro espacios vacíos.

- En la segunda línea (aquella que ha sido **indentada**) deberá contener la acción que se debe realizar si la condición es `True` (verdadera). En el ejemplo, se imprimiría `a` + `b`.

- Si la respuesta es `False`(falsa), no sucedería nada más.

Con todo lo anterior, las dos líneas de código se interpretarían como «Si `a` es igual a `b`, entonces se imprimirá `a + b`, de lo contrario no sucederá nada más»



#### 1.5.2 Declaración _else_ y _elif_

Como hemos visto, podemos pedir a **Python** que haga una acción si una condición es verdadera, pero ¿qué sucede si queremos que haga algo y la respuesta es falsa? En ese caso, solo deberemos incluir en nuestro programa la palabra reservada `else`.

Para entenderlo mejor, veamos un ejemplo similar al anterior:

```
if a == b:
    print(a + b)
else:
    print(b - a)    
```

Como has podido ver, se han ingresado dos líneas más:

- La tercera línea consta de la palabra reservada `else`, seguida de dos puntos (`:`), no deberá contener nada más que estos dos elementos.
- La cuarta línea al igual que la segunda está **indentada** y contiene otra instrucción, en nuestro ejemplo indica que si la condición en la primera línea es falsa, entonces mostrará las diferencia de `b` - `a`.

Por lo tanto estas líneas de código se interpretarían como «Si `a` es igual a `b`, entonces se imprimirá `a + b`, de lo contrario se imprimirá `b - a`».

Ahora que pasa si dentro de la **sentencia _if_** queremos meter otra **sentencia _if_**, algo así como preguntar que si algo no sucede entonces evaluará otra pregunta, vamos a verlo con un ejemplo.

In [ ]:
nombre = 'Ana'

if nombre == 'Sofia':
    print('Hola, Sofia.')
else:
    if nombre == 'Mariana':
        print('Hola, Mariana')
    else:
        print('Hola, persona no identificada.')

Hola, persona no identificada.


Como te darás cuenta debido a que hacemos una sentencia de condición dentro de otra sentencia de condición, debido a la **indentación**, el código se movió levemente a la derecha (a esto se le llamada **anidar**, cuando dentro de un código, hay otro código más), de hecho notarás que es complicado leer.

Su interpretación sería: «Si `nombre` es igual a `'Sofia'`, entonces se imprimirá `'Hola, Sofia.'`, de lo contrario se hará otra condición donde si `nombre` es igual a `Mariana`, entonces se imprimirá `Hola, Mariana`, de lo contrario se imprimirá `'Hola, persona no identificada.'`.

Para evitar **anidar** tanto y facilitar su lectura es cuando usamos la sentencia `elif` que es la combinación de las palabras _if_ y _else_, y como podrás imaginar junta las funciones de estas dos sentencias.

A continuación, veremos el mismo código de arriba, pero usando esta nueva palabra reservada:

In [ ]:
nombre = 'Ana'

if nombre == 'Sofia':
    print('Hola, Sofia.')
elif nombre == 'Mariana':
    print('Hola, Mariana')
else:
    print('Hola, persona no identificada.')

Hola, persona no identificada.


Es una estructura más simple de entender y más fácil de escribir, se han modificado las líneas de en medio.

- En lugar de usar `else` e `if` (en dos líneas distintas), se usa la palabra `elif` seguida de la nueva condición a evaluar y termina con dos puntos `:`

> **NOTA**: Recuerda poder dos puntos (`:`)...
> - después de la condición de procede a `if`,
> -  después de la condición de procede a `elif`,
> -  después de `else`,
> de lo contrario el programa indicará que existen errores.

El código de ejemplo se interpretaría como «Si la variable `nombre` es idéntica a la **cadena de texto** `'Sofia'`, entonces se muestra la **cadena de texto** `‘Hola, Sofia.’`, de lo contrario, si la variable `nombre` es idéntica a la **cadena de texto** `'Mariana'`, entonces se muestra la **cadena de texto** `‘Hola, Mariana.’`, de lo contrario se muestra la **cadena de texto** `‘Hola, persona no identificada.’`

### 1.5.3 Bucles

Un bucle no es más que un código que se repite cuantas veces nosotros se le señale. Un ejemplo de su uso es cuando, con una línea de código, queremos saludar a todo un grupo de personas, pero en lugar de escribir un saludo por persona pedimos que **Python** lo haga en automático.

Aunque existen varios tipos de bucles, nos centraremos en el **bucle for**, la cual hace que un elemento pase por todos los elementos de una lista. Lo anterior es más comprensible mediante un ejemplo:

In [ ]:
grupo = ['José', 'Arturo', 'Laura', 'Ulises', 'Fernanda']

for alumno in grupo:
    print('Buenos días, ' + alumno)

¿Pero qué ha sucedido? Lo explicaremos paso a paso, pero primero diremos cada elemento.

- Antes que nada, en nuestro ejemplo hemos definido una **lista** llamada `grupo` el cual contiene el nombre de cinco alumnos, pero desde la perspectiva de **Python** hablamos de una **lista** de **cadenas de texto**.
- Primero aparece la palabra reservada `for` que inicia el bucle.
- Luego sigue una palabra, la que gustes que es el iterable, en nuestro ejemplo usamos al palabra `alumno`.
- Continua con la palabra `in` que como lo dice su nombre es en donde se hará la iteración.
- Después es seguido ya sea por una **lista** o incluso un **rango**. En el ejemplo es la **lista** que hicimos al inicio (`grupo`).
- El siguiente paso es dar «enter» y como ya sabemos, aquello que va después de `:`necesita ser **indentada**.
- La última parte es escribir el código que queremos que se repita, donde veremos que está la palabra iterable.

Quizás hayas visto el código y no comprendiste qué sucedió, bueno explicaremos qué sucede.

La palabra `alumno` actúa como una variable, o sea que su valor cambia ¿y qué valores tendrá? Pues los que aparecen en la lista donde se itera, esto es en la lista `grupo`, el cual sabemos que tiene cinco valores: `['José', 'Arturo', 'Laura', 'Ulises', 'Fernanda']`.

El primer valor que tomará `alumno` es el primer elemento en la lista `grupo`, entonces `alumno` valdrá `'José'`, ahora en la segunda línea de código dice `print('Buenos días, ' + alumno)`, pero como sabemos que `alumno` tiene como valor `‘José’` es como si escribiéramos `print('Buenos días, ' + 'José').`

Una vez hecho esto, ahora regresamos a donde estaba el `for` y la palabra `alumno` valdrá el segundo valor de la **lista** `grupo`, o sea `‘Arturo’`. Y se volverá a ejecutar la segunda línea de código `print('Buenos días, ' + alumno)` pero ya sabemos que alumno ya no es el valor `'José'`, sino `‘Arturo’`. Por lo que es como si el código fuera `print('Buenos días, ' + 'Arturo')`.

Y lo anterior se repetirá cinco veces, una por cada elemento dentro de la **lista** `grupo`.

##### Ponte a prueba
Haz una lista llamada `amigos` donde deberás indicar el nombre de cinco amigos tuyos. Después, con un búcle deberás imprimir en pantalla un saludo a cada uno.

In [ ]:
amigos = ['Ian', 'Julieta', 'Erick', 'Cristina', 'Jorge']
for cuate in amigos:
  print('Holis ' + cuate)

#### Reto de la semana
Deberás hacer una lista llamada `estudiantes` con los siguientes nombres (cuidado, son cadenas de texto):
- Amanda
- Berenice
- Celia
- Danae
- Elena
- Fabiola
- Giselle

Ahora deberás hacer un **blucle for** en la que a cada nombre deberá comprobar si la última letra de cada nombre termina en `a`, si esa condición se cumple deberá decir:

`Amanda termina con la letra A`, y si no deberá decir: `Berenice no termina con la letra A`. Esto es para cada nombre en la lista.

Pista:
```
for i in lista_de_objetos:
    if i[1] == 'm':
        print(i + ' empieza con m')
    else:
        print(i + ' no empieza con m')
```


In [1]:
estudiante = ['Amanda', 'Berenice', 'Celia', ' Danae', 'Elena', 'Fabiola', 'Giselle']
for analisis in estudiante:
  if analisis[-1] == 'a':
    print(analisis + ' termina con a')
  else:
    print(analisis + ' no termina con a')

Amanda termina con a
Berenice no termina con a
Celia termina con a
 Danae no termina con a
Elena termina con a
Fabiola termina con a
Giselle no termina con a


## Mis próximos pasos

Hasta ahora has revisado algunas de las operaciones que Python permite ejecutar. Sabemos que al principio puede resultar algo complicado, pues estás aprendiendo un nuevo idioma; sin embargo, lo más importante es que sepas que cualquier persona puede aprender a programar con Python, pero debe ser paciente, disciplinada y probar constantemente sus avances.

¡Con el paso de los días te darás cuenta de todo lo que puedes hacer!

Revisa el contenido de la semana 2; así, en la próxima U Class nos enfocaremos en aplicar Python en el campo de la ciencia de datos.